libeos-updater-util/flatpak.c

```diff
@@ -0,0 +1,1411 @@
+  *out_action_type = (EosUpdaterUtilFlatpakRemoteRefActionType) enum_value->value;
+
+  return TRUE;
+}
+
+static gboolean
+parse_monotonic_counter (const gchar  *component,
+                         guint64      *out_counter,
+                         GError      **error)
+{
+  g_return_val_if_fail (out_counter != NULL, FALSE);
+
+  return eos_string_to_unsigned (component, 10, 0, G_MAXUINT32, out_counter, error);
+}
+
+static gchar *
+json_node_to_string (JsonNode *node)
+{
+  g_autoptr(JsonGenerator) gen = json_generator_new ();
+  json_generator_set_root (gen, node);
+
+  return json_generator_to_data (gen, NULL);
+}
+
+static gboolean
+parse_ref_kind (const gchar     *ref_kind_str,
+                FlatpakRefKind  *out_ref_kind,
+                GError         **error)
+{
+  FlatpakRefKind kind;
+
+  g_assert (out_ref_kind != NULL);
+
+  if (g_strcmp0 (ref_kind_str, "app") == 0)
+    {
+      kind = FLATPAK_REF_KIND_APP;
+    }
+  else if (g_strcmp0 (ref_kind_str, "runtime") == 0)
+    {
+      kind = FLATPAK_REF_KIND_RUNTIME;
+    }
+  else
+    {
+      g_set_error (error,
+                   EOS_UPDATER_ERROR,
+                   EOS_UPDATER_ERROR_MALFORMED_AUTOINSTALL_SPEC,
+                   "Invalid kind: %s", ref_kind_str);
+      return FALSE;
+    }
+
+  *out_ref_kind = kind;
+  return TRUE;
+}
+
+static gboolean
+parse_flatpak_ref_from_entry (JsonObject      *entry,
+                              const gchar    **out_app_name,
+                              FlatpakRefKind  *out_ref_kind,
+                              GError         **error)
+{
+  const gchar *app_name = NULL;
+  const gchar *ref_kind_str = NULL;
+  FlatpakRefKind kind;
+
+  g_return_val_if_fail (out_ref_kind != NULL, FALSE);
+  g_return_val_if_fail (out_app_name != NULL, FALSE);
+
+  app_name = json_object_get_string_member (entry, "app");
+
+  if (app_name == NULL)
+    {
+      g_set_error (error,
+                   EOS_UPDATER_ERROR,
+                   EOS_UPDATER_ERROR_MALFORMED_AUTOINSTALL_SPEC,
+                   "Expected an 'app' member");
+
+      return FALSE;
+    }
+
+  ref_kind_str = json_object_get_string_member (entry, "ref-kind");
+
+  if (ref_kind_str == NULL)
+    {
+      g_set_error (error,
+                   EOS_UPDATER_ERROR,
+                   EOS_UPDATER_ERROR_MALFORMED_AUTOINSTALL_SPEC,
+                   "Expected a 'ref-kind' member");
+      return FALSE;
+    }
+
+  if (!parse_ref_kind (ref_kind_str, &kind, error))
+    return FALSE;
+
+  *out_app_name = app_name;
+  *out_ref_kind = kind;
+
+  return TRUE;
+}
+
+static FlatpakRemoteRef *
+flatpak_remote_ref_from_install_action_entry (JsonObject *entry,
+                                              GError    **error)
+{
+  const gchar *app_name = NULL;
+  const gchar *collection_id = NULL;
+  FlatpakRefKind kind;
+
+  if (!parse_flatpak_ref_from_entry (entry, &app_name, &kind, error))
+    return NULL;
+
+  collection_id = json_object_get_string_member (entry, "collection-id");
```

Comment:
See comment below: use of `json_object_get_*_member()` will cause assertion failures if that member doesn’t exist or has the wrong type.